salt/transport/rabbitmq.py

```diff
@@ -0,0 +1,1243 @@
+        assert "transport_rabbitmq_auth" in self._opts
+        creds = self._opts["transport_rabbitmq_auth"]
+        assert "username" in creds and "password" in creds
+        self._creds = pika.PlainCredentials(creds["username"], creds["password"])
+
+
+        # TODO: figure out a better way to generate common queue names that are accessible to both master and minion,
+        # each possibly running on different machine
+        master_port = self._opts["master_port"] if "master_port" in self._opts else self._opts["ret_port"]
+        assert master_port
+
+        self._queue_name = "{}_{}".format(self._host, master_port)
+
+    @property
+    def queue_name(self):
+        return self._queue_name
+
+    def __del__(self):
+        try:
+            if self._channel and self._channel.is_open:
+                self._channel.close()
+        except:
+            pass
+
+        try:
+            if self._connection and self._connection.is_open:
+                self._connection.close()
+        except:
+            pass
+
+
+class RMQBlockingConnectionWrapper(RMQConnectionWrapperBase):
+    """
+    RMQConnection wrapper implemented that wraps a BlockingConnection.
+    Declares and binds a queue to a fanout exchange for publishing messages.
+    Caches connection and channel for reuse.
+    TODO: implement connection recovery and error handling (see rabbitmq docs for examples).
+    Not thread safe.
+    """
+
+    def __init__(self, opts):
+        super(RMQBlockingConnectionWrapper, self).__init__(opts)
+
+        self._connection = pika.BlockingConnection(pika.ConnectionParameters(host=self._host, credentials=self._creds))
+        self._channel = self._connection.channel()
+
+        # set up exchange for message broadcast
+        self._fanout_exchange_name = self.queue_name
+        self._channel.exchange_declare(exchange=self._fanout_exchange_name, exchange_type=ExchangeType.fanout)
+
+    def publish(self, payload, queue_name=None, reply_queue_name=None, correlation_id=None, auto_delete_queue=False, broadcast=True):
+        """
+        Publishes ``payload`` to the specified ``queue_name`` (via direct exchange or via fanout/broadcast),
+        passes along optional name of the reply queue in message metadata. Non-blocking.
+        Alternatively, broadcasts the ``payload`` to all bound queues (via fanout exchange).
+        :param correlation_id: correlation_id (used for the RPC pattern in conjunction with reply queue)
+        :param payload: message body
+        :param queue_name: queue name
+        :param reply_queue_name: optional name of the reply queue
+        :param auto_delete_queue: set the type of queue to "auto_delete" (
+        it will be deleted when the last consumer disappears)
+        :param broadcast: whether to broadcast via fanout exchange
+        :return:
+        """
+
+        publish_queue_name = queue_name if queue_name else self.queue_name
```

Comment:
**suggestion** in cases where empty/None values are to be skipped, like this one `publish_queue_name = queue_name or self.queue_name` is the preferred fallback approach.

Reply:
Done